alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,173 @@
+        let wait_notify = Box::new(on_exit);
+
+        let success = unsafe {
+            RegisterWaitForSingleObject(
+                &mut wait_handle,
+                child_handle,
+                Some(child_exit_callback::<F>),
+                Box::into_raw(wait_notify) as PVOID,
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if 0 == success {
+            Err(Error::last_os_error())
+        } else {
+            Ok(HandleWaitSignal { wait_handle: AtomicPtr::from(wait_handle) })
+        }
+    }
+}
+
+impl Drop for HandleWaitSignal {
+    fn drop(&mut self) {
+        unsafe {
+            // Cancel WinAPI wait for child process exiting
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
+        }
+    }
+}
+
+pub struct ChildProcessWatcher {
+    _on_exit: HandleWaitSignal,
+    events_rx: Receiver<ChildEvent>,
+}
+
+impl ChildProcessWatcher {
+    pub fn new(subprocess_handle: HANDLE) -> Result<ChildProcessWatcher, Error> {
+        let (sender, receiver) = channel();
+        let on_exit = HandleWaitSignal::new(subprocess_handle, move || {
+            let _ = sender.send(ChildEvent::Exited);
+        })?;
+
+        Ok(ChildProcessWatcher { _on_exit: on_exit, events_rx: receiver })
+    }
+
+    pub fn events_rx(&self) -> &Receiver<ChildEvent> {
+        &self.events_rx
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::io::Error;
+    use std::ptr;
+    use std::time::Duration;
+
+    use std::sync::mpsc::channel;
+
+    use widestring::U16CString;
+
+    use winapi::shared::ntdef::LPWSTR;
+    use winapi::um::processthreadsapi::{
+        CreateProcessW, TerminateProcess, PROCESS_INFORMATION, STARTUPINFOW,
+    };
+
+    use super::*;
+
+    fn make_cmd_process() -> Result<HANDLE, Error> {
+        let mut pi: PROCESS_INFORMATION = Default::default();
+        let mut si: STARTUPINFOW = Default::default();
+        let cmdline = U16CString::from_str("cmd.exe").unwrap();
+
+        unsafe {
+            if 0 == CreateProcessW(
+                ptr::null(),
+                cmdline.as_ptr() as LPWSTR,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                false as i32,
+                Default::default(),
+                ptr::null_mut(),
+                ptr::null_mut(),
+                &mut si as *mut STARTUPINFOW,
+                &mut pi as *mut PROCESS_INFORMATION,
+            ) {
+                Err(Error::last_os_error())
+            } else {
+                Ok(pi.hProcess)
+            }
+        }
+    }
+
+    #[test]
+    pub fn on_handle_wait_runs_callback_when_process_exits() {
+        // The test should only take few milliseconds to succeed but will
+        // only fail due to timeout. It can be safely assumed that if callback
+        // has not been called after `WAIT_TIMEOUT` duration, there is some
+        // issue with the code under test.
```

Comment:
I'm still a bit skeptical about this comment. Saying it 'should only take a few milliseconds' seems mostly like a guess, which usually isn't super great. We should instead determine why this could potentially timeout and document this specific case, if we can't determine reliably how long a test will take, we'll have test failures left and right which is terrible for our trust in the test suite.

This sounds a lot worse than it is though, since we already know why we need the timeout and what the potential problems are. So we just need to clarify the comment and talk about exactly what specific things can happen and why this timeout resolves them, rather than having a vague statement about what could be going on.

Additionally this comment kinda makes it sound like the timeout would only happen when the subprocess fails and is unrelated to us. However the timeout is exactly what we use to determine if **our** functionality is correct. So a timeout would also happen if we break our exit handling.

So I'm thinking something more like this:

> We use a timeout to determine a maximum timeframe in which the exit event has to be emitted, since failure will not emit anything.

Reply:
Should I just remove it completely?

Reply:
I think it's useful to point out that the timeout is used to determine success of the test, kinda like the example I've provided.

Reply:
I replaced it with [a comment on the actual line](https://github.com/cfiet/alacritty/blob/fix/winpty-freeze-on-termination/alacritty_terminal/src/tty/windows/child.rs#L132) that does the waiting, I feel it's a bit more clear there.

Reply:
Yeah, was considering to suggest that myself. Great thinking.